src/imputors/interp.jl

```diff
@@ -81,10 +69,41 @@ function _impute!(data::AbstractVector{<:Union{T, Missing}}, imp::Interpolate) w
     return data
 end
 
-# Default cases where no rounding behaviour is specified
-_calculate_increment(a, b, n, ::Nothing) = (b - a) / n
-_calculate_increment(a::Unsigned, b::Unsigned, n, r::Nothing) = _calculate_increment(Int(a), Int(b), n, r)
+"""
+Set a vector slice over the values of a generator, starting from `after+1`
+"""
+function _gen_set!(v::AbstractVector, after::Integer, gen)
```

Comment:
I think we're making this more complicated than it needs to be. The only special cases we need are:

1. `_calculate_increment` for  `Unsigned`.
2. `round` on insert if a rounding mode is specified.

Reply:
Yeah definitely could be simplified. I simplified it a little, but your last push looks simpler. However I ran some horse races and using a generator wrapped in a function seems to consistently (~1k, ~500k, ~3mil) be about 20% faster. I'm not exactly sure why, but I found that when benchmarking. Something related to compiler optimzations...

Reply:
I boiled it down to this. Of course, `_calculate_increment` can be put at the top level. I think I did it this way to try to squeeze out some extra performance lol. Anyway, that has a negligible impact but the generator seems to do something.
```
function _gen_set!(v::AbstractVector, after::Integer, gen)
    for (i, val) in enumerate(gen)
       v[after+i] = val
    end
end

function _gen_interp(a, b, n, r)
    inc = _calculate_increment(a, b, n)
    (a + inc*i for i=1:n)
end

function _gen_interp(a::T, b::T, n, r::RoundingMode) where {T<:Integer}
    inc = _calculate_increment(a, b, n)
    (round(T, a + inc*i, r) for i=1:n)
end

_calculate_increment(a, b, n) = (b - a) / n
_calculate_increment(a::T, b::T, n) where {T<:Unsigned} = _calculate_increment(Int(a), Int(b), n)
```

Reply:
So this is as simple as I got it down to:
```
               ...
if imp.limit === nothing || gap_sz <= imp.limit
    inc = _calculate_increment(data[prev_idx], data[next_idx], n)
    gen = _gen_interp(data[prev_idx], inc, gap_sz+1, imp.r)
    _gen_set!(data, prev_idx, gen)
end
               ...
function _gen_set!(v::AbstractVector, after::Integer, gen)
    for (i, val) in enumerate(gen)
       v[after+i] = val
    end
end

_gen_interp(a, inc, n, r) = (a + inc*i for i=1:n)
_gen_interp(a::T, inc, n, r::RoundingMode) where {T<:Integer} = (round(T, a + inc*i, r) for i=1:n)

_calculate_increment(a, b, n) = (b - a) / n
_calculate_increment(a::T, b::T, n) where {T<:Unsigned} = _calculate_increment(Int(a), Int(b), n)
```

Reply:
Cool, thanks for running those benchmarks! I wouldn’t have expected that. I might take a look at what inference is doing before merging. It’s possible the function boundary avoids having inference give up in the loop or something. Which version of Julia were you running?

Reply:
Good idea, yes I would like to know too. I'm running version 1.10.2

Reply:
Alright, I took a look, and you're correct. There's a noticeable improvement for arrays of integers. It appears that the original code has type instability in the loop where we're rounding the final value.

- `_impute!` is the base function
- `_impute2!` is your generator solution


ints
```
julia> @benchmark Impute._impute!(data, imp) setup=(data=Impute.trycopy(x)) evals=100 seconds=20
BenchmarkTools.Trial: 10000 samples with 100 evaluations.
 Range (min … max):  7.448 μs …  35.280 μs  ┊ GC (min … max): 0.00% … 77.67%
 Time  (median):     7.492 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   7.591 μs ± 831.703 ns  ┊ GC (mean ± σ):  0.63% ±  3.69%

  ▇█▇▆▆▅▄▃▃▂▁▁▁▁   ▃▂▂▂▂▁▁          ▁                         ▂
  ███████████████████████████▇▆▆▇▇▆▇██▇▆▇▅▅▆▆▅▆▆▄▃▄▃▄▄▄▄▃▄▄▅▂ █
  7.45 μs      Histogram: log(frequency) by time      8.28 μs <

 Memory estimate: 1.91 KiB, allocs estimate: 122.
```
vs
```
julia> @benchmark Impute._impute2!(data, imp) setup=(data=Impute.trycopy(x)) evals=100 seconds=20
BenchmarkTools.Trial: 10000 samples with 100 evaluations.
 Range (min … max):  6.183 μs …   8.882 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     6.248 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   6.270 μs ± 116.534 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▄▄▅▄▅█▆▅▅▄▄▅▃▃▂▁▁▁▁▁ ▁▁  ▁                                 ▂
  ████████████████████████████████▇▇▆▆▆▅▅▆▆▆▆▆▇▇▅▆▆▆▅▅▅▅▅▅▅▅▄ █
  6.18 μs      Histogram: log(frequency) by time      6.75 μs <

 Memory estimate: 1 bytes, allocs estimate: 0.
```

But for floats, the original implementation appears to work better.
```
julia> @benchmark Impute._impute!(data, imp) setup=(data=Impute.trycopy(x)) evals=100 seconds=20
BenchmarkTools.Trial: 10000 samples with 100 evaluations.
 Range (min … max):  4.735 μs …   5.951 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     4.790 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   4.822 μs ± 102.945 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▃▄▄▄▃▆█▄▅▅▄▃▅▃▃▂▂▃▃▂▂▁▂▂▃▃▂▂▂▂▂      ▂▁      ▂             ▂
  █████████████████████████████████████▇██▇▇▇▇▇██▆▇▅▇▆▇▆▁▅▅▅▆ █
  4.73 μs      Histogram: log(frequency) by time      5.19 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.
```
vs
```
julia> @benchmark Impute._impute2!(data, imp) setup=(data=Impute.trycopy(x)) evals=100 seconds=20
BenchmarkTools.Trial: 10000 samples with 100 evaluations.
 Range (min … max):  6.147 μs …   8.098 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     6.238 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   6.288 μs ± 155.297 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▆     █
  █▅▅▄▅▃█▅▅▄▅▄▂▂▂▃▂▂▂▂▂▃▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  6.15 μs         Histogram: frequency by time        6.92 μs <

 Memory estimate: 1 bytes, allocs estimate: 0.
```

However, if we simply move the loop into a separate function, then we remove the type instability and avoid the additional cost of the generator.

ints
```
julia> @benchmark Impute._impute3!(data, imp) setup=(data=Impute.trycopy(x)) evals=100 seconds=20
BenchmarkTools.Trial: 10000 samples with 100 evaluations.
 Range (min … max):  6.177 μs …   9.223 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     6.247 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   6.295 μs ± 178.248 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █    ▇▁
  █▅▄▄▃██▅▆▄▅▃▂▂▂▁▂▃▂▂▂▂▃▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  6.18 μs         Histogram: frequency by time        6.85 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.
```
floats
```
julia> @benchmark Impute._impute3!(data, imp) setup=(data=Impute.trycopy(x)) evals=100 seconds=20
BenchmarkTools.Trial: 10000 samples with 100 evaluations.
 Range (min … max):  4.673 μs …  6.104 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     4.678 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   4.718 μs ± 90.578 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▃▄▄▃▃▃▅▃▁▁        ▃  ▁▁                                   ▁
  ███████████████▇▇▇▆██▇███▇▇▇▆▆▆▆▅▅▅▅▅▅▄█▅▄▄▄▄▃▅▄▄▄▄▄▄▅▄▄▃▄ █
  4.67 μs      Histogram: log(frequency) by time     5.11 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.
```
I'll make the changes and update.

Reply:
Good to hear, sounds like that is the way to go!